main.go

```diff
@@ -58,10 +59,31 @@ func (lf logFunc) Println(v ...interface{}) {
 	lf("msg", fmt.Sprintln(v...))
 }
 
+func decompressGzip() func(http.Handler) http.Handler {
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			contentEncoding := r.Header.Get("Content-Encoding")
+			if strings.Contains(contentEncoding, "gzip") {
+				gr, err := gzip.NewReader(r.Body)
+				if err != nil {
+					w.WriteHeader(http.StatusBadRequest)
+					w.Write([]byte(err.Error()))
+				} else {
+					r.Body = gr
+					h.ServeHTTP(w, r)
+				}
+			} else {
+				h.ServeHTTP(w, r)
+			}
+		})
+	}
+}
```

Comment:
Following thoughts:

- I think this doesn't have to be a doubly nested function. The outermost layer doesn't do anything.
- `ServeHTTP` won't call `Close` but someone has to close the gzip reader.
- I think we should only try to unzip if the Content-Encoding header is an exact match (who knows which future compression algorithms will contain "gzip").
- There is `http.Error` to write an error to a `ResponseWriter`.
- General style concern: Go tends towards avoiding the indented `else` branch if you can avoid it with a `return`.
- Finally, I'm thinking the name should better express that this is about request compression, not response compression. (Others might suffer the same confusion I did.) `gunzipRequest` perhaps?

In sum, this is my proposal:
```Go
func gunzipRequest(h http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		contentEncoding := r.Header.Get("Content-Encoding")
		if contentEncoding == "gzip" {
			gr, err := gzip.NewReader(r.Body)
			if err != nil {
				http.Error(w, err.Error(), http.StatusBadRequest)
				return
			}
			defer gr.Close()
			r.Body = gr
		}
		h.ServeHTTP(w, r)
	})
}
```

Reply:
this code is more simple and clearly